File: json/json-annotation-processor/src/main/java/ru/tinkoff/kora/json/annotation/processor/reader/SealedInterfaceReaderGenerator.java
package ru.tinkoff.kora.json.annotation.processor.reader;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.SealedTypeUtils;
import ru.tinkoff.kora.json.annotation.processor.JsonTypes;
import ru.tinkoff.kora.json.annotation.processor.JsonUtils;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.io.IOException;
import java.util.List;

import static ru.tinkoff.kora.annotation.processor.common.CommonUtils.decapitalize;

public class SealedInterfaceReaderGenerator {
    private final Types types;
    private final Elements elements;

    public SealedInterfaceReaderGenerator(ProcessingEnvironment processingEnvironment) {
        this.types = processingEnvironment.getTypeUtils();
        this.elements = processingEnvironment.getElementUtils();
    }

    public TypeSpec generateSealedReader(TypeElement jsonElement) {
        var typeName = JsonUtils.jsonReaderName(jsonElement);
        var typeBuilder = TypeSpec.classBuilder(typeName)
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.koraGenerated)
                .addMember("value", CodeBlock.of("$S", SealedInterfaceReaderGenerator.class.getCanonicalName()))
                .build())
            .addSuperinterface(ParameterizedTypeName.get(JsonTypes.jsonReader, ClassName.get(jsonElement)))
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addOriginatingElement(jsonElement);

        for (var typeParameter : jsonElement.getTypeParameters()) {
            typeBuilder.addTypeVariable(TypeVariableName.get(typeParameter));
        }
        var permittedSubclasses = SealedTypeUtils.collectFinalPermittedSubtypes(this.types, this.elements, jsonElement);

        this.addReaders(typeBuilder, permittedSubclasses);

        var discriminatorField = JsonUtils.discriminatorField(this.types, jsonElement);
        if (discriminatorField == null) {
            discriminatorField = "@type";
        }
        var method = MethodSpec.methodBuilder("read")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addException(IOException.class)
            .addParameter(JsonTypes.jsonParser, "_parser")
            .returns(ClassName.get(jsonElement))
            .addAnnotation(Override.class)
            .addAnnotation(Nullable.class);
        method.addCode("var bufferingParser = new $T(_parser);\n", JsonTypes.bufferingJsonParser);
        method.addCode("var discriminator = $T.readStringDiscriminator(bufferingParser, $S);\n", JsonTypes.discriminatorHelper, discriminatorField);
        method.addCode("if (discriminator == null) throw new $T(_parser, $S);\n", JsonTypes.jsonParseException, "Discriminator required, but not provided");
        method.addCode("var bufferedParser = $T.createFlattened(false, bufferingParser.reset(), _parser);\n", JsonTypes.jsonParserSequence);
        method.addCode("bufferedParser.nextToken();\n");
        method.addCode("return switch(discriminator) {$>\n");
        for (var elem : permittedSubclasses) {
            var readerName = getReaderFieldName(elem);
            var discriminatorValues = JsonUtils.discriminatorValue(elem);
            for (var discriminatorValue : discriminatorValues) {
                method.addCode("case $S -> $L.read(bufferedParser);\n", discriminatorValue, readerName);
            }
        }
        method.addCode("default -> throw new $T(_parser, $S + discriminator + \"'\");$<\n};", JsonTypes.jsonParseException, "Unknown discriminator: '");
        typeBuilder.addMethod(method.build());

        return typeBuilder.build();
    }

    private void addReaders(TypeSpec.Builder typeBuilder, List<? extends Element> jsonElements) {
        var constructor = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        jsonElements.forEach(elem -> {
            var fieldName = getReaderFieldName(elem);
            var fieldType = ParameterizedTypeName.get(JsonTypes.jsonReader, TypeName.get(elem.asType()));
            var readerField = FieldSpec.builder(fieldType, fieldName, Modifier.PRIVATE, Modifier.FINAL);
            constructor.addParameter(fieldType, fieldName);
            constructor.addStatement("this.$L = $L", fieldName, fieldName);
            typeBuilder.addField(readerField.build());
        });
        typeBuilder.addMethod(constructor.build());
    }

    @Nonnull
    private String getReaderFieldName(Element elem) {
        return decapitalize(elem.getSimpleName().toString() + "Reader");
    }
}

File: json/json-symbol-processor/src/main/kotlin/ru/tinkoff/kora/json/ksp/reader/SealedInterfaceReaderGenerator.kt
package ru.tinkoff.kora.json.ksp.reader

import com.fasterxml.jackson.core.JsonParseException
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSDeclaration
import com.google.devtools.ksp.symbol.KSTypeParameter
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.addOriginatingKSFile
import com.squareup.kotlinpoet.ksp.toClassName
import ru.tinkoff.kora.json.ksp.JsonTypes
import ru.tinkoff.kora.json.ksp.detectSealedHierarchyTypeVariables
import ru.tinkoff.kora.json.ksp.discriminatorField
import ru.tinkoff.kora.json.ksp.jsonReaderName
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValue
import ru.tinkoff.kora.ksp.common.KspCommonUtils.collectFinalSealedSubtypes
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated
import ru.tinkoff.kora.ksp.common.KspCommonUtils.toTypeName
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import java.util.*

class SealedInterfaceReaderGenerator(private val resolver: Resolver, logger: KSPLogger) {
    fun generateSealedReader(jsonClassDeclaration: KSClassDeclaration): TypeSpec {
        val subclasses = jsonClassDeclaration.collectFinalSealedSubtypes().toList()
        val (typeArgMap, readerTypeVariables) = detectSealedHierarchyTypeVariables(jsonClassDeclaration, subclasses)

        val typeName = if (jsonClassDeclaration.typeParameters.isEmpty())
            jsonClassDeclaration.toClassName() else
            jsonClassDeclaration.toClassName().parameterizedBy(readerTypeVariables)

        val readerInterface = JsonTypes.jsonReader.parameterizedBy(typeName)

        val typeBuilder = TypeSpec.classBuilder(jsonClassDeclaration.jsonReaderName())
            .generated(SealedInterfaceReaderGenerator::class)
            .addSuperinterface(readerInterface)
            .addModifiers(KModifier.PUBLIC)
        jsonClassDeclaration.containingFile?.let { typeBuilder.addOriginatingKSFile(it) }

        readerTypeVariables.forEach {
            if (it is TypeVariableName) {
                typeBuilder.addTypeVariable(it)
            }
        }

        addReaders(typeBuilder, subclasses, typeArgMap)

        val discriminatorField = jsonClassDeclaration.discriminatorField(resolver)
            ?: throw ProcessingErrorException("Sealed interface should have @JsonDiscriminatorField annotation", jsonClassDeclaration)
        val function = FunSpec.builder("read")
            .addModifiers(KModifier.PUBLIC, KModifier.OVERRIDE)
            .addParameter("_parser", JsonTypes.jsonParser)
            .returns(typeName.copy(nullable = true))
        function.addCode("val bufferedParser = %T(_parser)\n", JsonTypes.bufferedParserWithDiscriminator)
        function.addCode("val discriminator = bufferedParser.getDiscriminator(%S) ", discriminatorField)
        function.addCode("?: throw %T(_parser, %S)\n", JsonTypes.jsonParseException, "Discriminator required, but not provided")
        function.addCode("bufferedParser.resetPosition()\n")
        function.beginControlFlow("return when(discriminator) {")
        subclasses.forEach { elem ->
            val readerName = getReaderFieldName(elem)
            val requiredDiscriminatorValue = elem.findAnnotation(JsonTypes.jsonDiscriminatorValue)
                ?.findValue<String>("value")
                ?: elem.simpleName.asString()
            function.addCode(
                "%S -> %L.read(bufferedParser)\n",
                requiredDiscriminatorValue,
                readerName
            )
        }
        function.addCode("else -> throw %T(_parser, %S)", JsonParseException::class.java, "Unknown discriminator")
        function.endControlFlow()
        typeBuilder.addFunction(function.build())
        return typeBuilder.build()
    }


    private fun addReaders(typeBuilder: TypeSpec.Builder, jsonElements: List<KSClassDeclaration>, typeArgMap: IdentityHashMap<KSTypeParameter, TypeName>) {
        val constructor = FunSpec.constructorBuilder()
        jsonElements.forEach { sealedSub ->
            val fieldName = getReaderFieldName(sealedSub)
            val subtypeTypeName = sealedSub.toTypeName(sealedSub.typeParameters.map { typeArgMap[it] ?: STAR })
            val fieldType = JsonTypes.jsonReader.parameterizedBy(subtypeTypeName)
            val readerField = PropertySpec.builder(fieldName, fieldType, KModifier.PRIVATE)
            constructor.addParameter(fieldName, fieldType)
            constructor.addStatement("this.%L = %L", fieldName, fieldName)
            typeBuilder.addProperty(readerField.build())
        }
        typeBuilder.primaryConstructor(constructor.build())
    }

    private fun getReaderFieldName(elem: KSDeclaration): String {
        return elem.simpleName.asString().replaceFirstChar { it.lowercaseChar() } + "Reader"
    }
}
